interpreter/types/object.go

```diff
@@ -0,0 +1,165 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Adapter package defines utilities for adapting plain Go structs into
+// structs suitable for consumption with CEL.
+package types
+
+import (
+	"fmt"
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/any"
+	"github.com/google/cel-go/interpreter/types/aspects"
+	"reflect"
+)
+
+// ObjectValue wraps accesses to protobuf messages in order to ensure
+// appropriate type conversions between Go and CEL.
+//
+// ObjectValue values are comparable, iterable, and support indexed access by
+// field name.
+//
+// Note: iterators will iterate over all default and non-default fields, but
+// for oneof field sets will only include the field name of the non-default
+// oneof.
+type ObjectValue interface {
+	aspects.Equaler
+	aspects.Indexer
+	aspects.Iterable
+
+	// Value of the underlying message.
+	Value() interface{}
+}
+
+type protoValue struct {
+	value       interface{}
+	refValue    *reflect.Value
+	isAny       bool
+	fieldValues map[string]interface{}
+}
+
+func NewProtoValue(value interface{}) ObjectValue {
+	// Unwrap any.Any values on construction.
+	_, isAny := value.(*any.Any)
+	if isAny {
+		// TODO: check whether the any being null is valid.
+		dynAny := &ptypes.DynamicAny{}
+		if err := ptypes.UnmarshalAny(value.(*any.Any), dynAny); err != nil {
+			panic(err)
```

Comment:
Generally should not panic.  Suggest adding error return parameter to this function instead.

Reply:
I've added a TODO. Agree that this should be handled differently, but wasn't sure how a missing type should have been represented until very recently.